trace/trace.go

```diff
@@ -0,0 +1,348 @@
+// field will be summed across all spans and added to the trace as a total. It
+// is especially useful for doing things like adding the duration spent talking
+// to a specific external service - eg database time. The root span will then
+// get a field that represents the total time spent talking to the database from
+// all of the spans that are part of the trace.
+func (s *Span) AddRollupField(key string, val float64) {
+	s.rollupLock.Lock()
+	if s.rollupFields != nil {
+		s.rollupFields[key] += val
+	}
+	s.rollupLock.Unlock()
+	if s.trace != nil {
+		s.trace.rollupLock.Lock()
+		defer s.trace.rollupLock.Unlock()
+		s.trace.rollupFields[key] += val
+	}
+}
+
+// AddTraceField adds a key/value pair to this span and all others involved in
+// this trace. These fields are also passed along to downstream services. This
+// method is functionally identical to `Trace.AddField()`.
+func (s *Span) AddTraceField(key string, val interface{}) {
+	// these get added to this span when it gets sent, so don't bother adding
+	// them here
+	if s.trace != nil {
+		s.trace.AddField(key, val)
+	}
+}
+
+// Finish marks a span complete. It does not actually send the span to
+// Honeycomb; spans stick around until the entire trace is complete and then get
+// dispatched to Honeycomb. Finishing a span also triggers finishing all
+// synchronous child spans - in other words, if any synchronous child span has
+// not yet been finished, finishing the parent will finish the children as well.
+func (s *Span) Finish() {
+	if s.ev == nil {
+		return
+	}
+	// finish the timer for this span
+	if s.timer != nil {
+		dur := s.timer.Finish()
+		s.ev.AddField("duration_ms", dur)
+	}
+	// set trace IDs for this span
+	s.ev.AddField("trace.trace_id", s.trace.traceID)
+	if s.parentID != "" {
+		s.ev.AddField("trace.parent_id", s.parentID)
+	}
+	s.ev.AddField("trace.span_id", s.spanID)
+	// add rollup fields to the event
+	for k, v := range s.rollupFields {
+		s.ev.AddField(k, v)
+	}
+	for _, child := range s.children {
+		if !child.AmAsync() {
+			if !child.amFinished {
+				child.AddField("meta.finished_by_parent", true)
+				child.Finish()
+			}
+		}
+	}
+	s.amFinished = true
+	// if we're closing the root span, send the whole trace
+	if s.amRoot {
+		s.trace.Send()
+	}
+}
+
+// AmAsync reveals whether the span is asynchronous (true) or synchronous (false).
+func (s *Span) AmAsync() bool {
+	return s.amAsync
+}
+
+// GetChildren returns a list of all child spans (both synchronous and
+// asynchronous).
+func (s *Span) GetChildren() []*Span {
+	return s.children
+}
+
+// Get Parent returns this span's parent.
+func (s *Span) GetParent() *Span {
+	return s.parent
+}
+
+// ChildAsyncSpan creates a child of the current span that is expected to
+// outlive the current span (and trace). Async spans must be manually sent using
+// the `Send()` method but are otherwise identical to normal spans.
+func (s *Span) ChildAsyncSpan(ctx context.Context) (context.Context, *Span) {
```

Comment:
Nathan objected to `New` in the name because it has strong meaning in normal go usage as a constructor (which this really isn't in the way that it would be in other Go things)